crates/ruff_python_ast/src/name.rs

```diff
@@ -133,15 +124,126 @@ impl<'a> CallPathBuilder<'a> {
         self.segments.extend_from_slice(segments);
     }
 
-    pub fn build(self) -> CallPath<'a> {
-        CallPath {
+    pub fn build(self) -> QualifiedName<'a> {
+        QualifiedName {
             segments: self.segments,
         }
     }
 }
 
-/// Convert an `Expr` to its [`CallPath`] segments (like `["typing", "List"]`).
-fn collect_call_path(expr: &Expr) -> Option<SmallVec<[&str; 8]>> {
+impl Display for QualifiedName<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        format_call_path_segments(self.segments(), f)
+    }
+}
+
+/// Convert an `Expr` to its call path (like `List`, or `typing.List`).
+pub fn compose_call_path(expr: &Expr) -> Option<String> {
```

Comment:
(Do these need to be `pub`? Are there usages that don't go through the structs?)

Reply:
There are :( I can try to eliminate it entirely because it's really just a `UnqualifiedName::from_expr(expr).map(ToString::to_string)` and this doesm't seem a good pattern IMO